### DISCUSSION TITLE
Incremental analysis

### DISCUSSION THREAD

**Discussion by ADD-Eugenio-Lopez** [dotnet/roslyn#76485] [Q&A] [answered]
Is there a way to do incremental code analysis? That is, the new code is analyzed and only the new issues are shown.
This possibility of code analysis would be very interesting.

**Comment by CyrusNajmabadi** [comment, marked as answer]
Yes. This is how Roslyn analyzers already work.

**Comment by ADD-Eugenio-Lopez** [comment]
Whenever I run the analysis on a project in Visual Studio I get the issues on all the solution code, if it really works like this, please could you explain how to get it or where the documentation is that explains it?

**Reply by CyrusNajmabadi** [reply]
As examples, when you edit a method body, we only renanalyze the code in that method body.  Similarly, if the only changes made are private changes, we don't reanalyze downstream projects.  Similarly, if you make public changes, we do analyze downstream projects, but not projects that are unaffected. etc. etc.

> Whenever I run the analysis on a project in Visual Studio I get the issues on all the solution code,

You seem to be referring to the results that are presented.  The results are always shown as if they are 'complete'.  'incremental analysis' is a mechanism to improve performance (by not recomputing things that would not have changed).  it doesn't end up changing the final results computed.

**Reply by ADD-Eugenio-Lopez** [reply]
I understand that default behavior is incremental, I'm going to rephrase the initial question.

Could the results of the analysis be saved in a file and obtain the results of the changes with respect to that saved file?

**Reply by CyrusNajmabadi** [reply]
Sure. I imagine that is something that could be done.

**Reply by 333fred** [reply]
But, to be clear, it is not something that Roslyn supports today, and not something that we're interested in implementing for either VS or VS Code.

**Reply by CyrusNajmabadi** [reply]
Note: the easiest way to accomplish this yourself would be to just have roslyn run as a library in some command line tool you create.  Have it analyze everything, and write all those analysis results to a file.  Then do the same in the future.  But instead of writing out everything, do a diff between what you are producing now and what was produced the last time, and only write out the diff.

**Reply by ADD-Eugenio-Lopez** [reply]
Thanks for all.
